File: Cocoa/Plug-Ins/Template Editor/ElementRSID.m
#import "ElementRSID.h"
#import "ElementCASE.h"
#import "Resource.h"
#import "ResourceDocument.h"
#import "ResourceDataSource.h"
#import "ResKnifeResourceProtocol.h"

@implementation ElementRSID
@synthesize resType = _resType;

- (instancetype)initForType:(NSString *)t withLabel:(NSString *)l
{
    if (self = [super initForType:t withLabel:l]) {
        /*
         * Determine resource type and offset from label
         * Resource type is a 4-char code enclosed in single (or smart) quotes
         * Offset is a number followed by a +
         * E.g. "Extension scope info 'scop' -27136 +"
         */
        NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"(?:.*[‘'](.{4})['’])?(?:.*?(-?[0-9]+) *[+])?" options:0 error:nil];
        NSTextCheckingResult *result = [regex firstMatchInString:l options:0 range:NSMakeRange(0, l.length)];
        NSRange r = [result rangeAtIndex:1];
        if (r.location != NSNotFound) {
            _resType = GetOSTypeFromNSString([l substringWithRange:[result rangeAtIndex:1]]);
        }
        r = [result rangeAtIndex:2];
        if (r.location != NSNotFound) {
            _offset = (short)[[l substringWithRange:r] intValue];
        }
    }
    return self;
}

- (OSType)resType
{
    return _resType;
}

- (void)setResType:(OSType)resType
{
    _resType = resType;
    [self loadCases];
}

- (void)readSubElements
{
    if (!self.resType) {
        // See if we can bind to a preceding TNAM field
        Element *tnam = [self.parentList previousOfType:@"TNAM"];
        if (!tnam) {
            NSLog(@"Could not determine resource type for RSID.");
            [super readSubElements];
            return;
        }
        [self bind:@"resType" toObject:tnam withKeyPath:@"tnam" options:nil];
    }
    self.fixedCases = [NSMutableArray new];
    Element *element = [self.parentList peek:1];
    while (element.class == ElementCASE.class) {
        [self.fixedCases addObject:[self.parentList pop]];
        element = [self.parentList peek:1];
    }
    self.caseMap = [NSMutableDictionary new];
    [self loadCases];
}

- (void)loadCases
{
    NSMutableArray *cases = [NSMutableArray new];
    [self.caseMap removeAllObjects];
    for (ElementCASE *element in self.fixedCases) {
        NSString *option = [NSString stringWithFormat:@"%@ = %@", element.symbol, element.value];
        NSString *display = [NSString stringWithFormat:@"%@ = %@", element.value, element.symbol];
        [cases addObject:option];
        [self.caseMap setObject:display forKey:element.value];
    }
    if (self.resType) {
        // Find resources in all documents
        NSMutableArray *resCases = [NSMutableArray new];
        for (NSDocument *doc in [[NSDocumentController sharedDocumentController] documents]) {
            NSArray *resources = [[(ResourceDocument *)doc dataSource] allResourcesOfType:self.resType];
            for (id <ResKnifeResource> resource in resources) {
                if (!resource.name.length) continue; // No point showing resources with no name
                NSString *resID = [@(resource.resID-self.offset) stringValue];
                if (![self.caseMap valueForKey:resID]) {
                    NSString *idDisplay = [self resIDDisplay:resID];
                    [resCases addObject:[NSString stringWithFormat:@"%@ = %@", resource.name, idDisplay]];
                    [self.caseMap setValue:[NSString stringWithFormat:@"%@ = %@", idDisplay, resource.name] forKey:resID];
                }
            }
        }
        // Sort the resources by name before adding them to the list
        [cases addObjectsFromArray:[resCases sortedArrayUsingSelector:@selector(caseInsensitiveCompare:)]];
    }
    self.cases = cases; // This triggers the combo box refresh
}

- (NSString *)resIDDisplay:(NSString *)resID
{
    // If an offset is used, the value will be displayed as "value (#actual id)"
    return self.offset ? [resID stringByAppendingFormat:@" (#%d)", [resID intValue]+self.offset] : resID;
}

- (id)transformedValue:(id)value
{
    value = [value stringValue];
    return [self.caseMap valueForKey:value] ?: [self resIDDisplay:value];
}

- (id)reverseTransformedValue:(id)value
{
    value = [[value componentsSeparatedByString:@" = "] lastObject];
    if (self.offset) value = [[value componentsSeparatedByString:@" "] firstObject];
    return value ?: @"";
}

@end

File: ResForge/Template Editor/Elements/ElementRSID.swift
import AppKit
import OrderedCollections
import RFSupport

/*
 * RSID allows selecting a resource id of a given type
 * The parameters are specified in the label
 * Resource type is a 4-char code enclosed in single (or smart) quotes
 * Offset is a number followed by a + (a value of zero refers to this id)
 * Limit is a number immediately following the +
 * If limit is specified the list will only show resources between offset and offset+limit
 * E.g. "Extension scope info 'scop' -27136 +2" will show 'scop' resources between -27136 and -27134
 * If the resource type cannot be determined from the label, it will look for a preceding TNAM element to determine the type
 *
 * Implements RSID, LRID
 */
class ElementRSID<T: FixedWidthInteger & SignedInteger>: CasedElement, LinkingComboBoxDelegate {
    var tValue: T = 0 {
        didSet {
            self.updateLinkIcon()
        }
    }
    @objc dynamic private var value: NSNumber {
        get { tValue as! NSNumber }
        set { tValue = newValue as! T }
    }
    @objc private var resType = "" {
        didSet {
            self.loadCases()
        }
    }
    private var offset: Int = 0
    private var range: ClosedRange<Int>?
    private var fixedMap: OrderedDictionary<AnyHashable, ElementCASE> = [:]
    private(set) var linkIcon: NSImage.Name?

    deinit {
        self.unbind(NSBindingName("resType"))
    }

    override func configure() throws {
        // Determine parameters from label
        let regex = try! NSRegularExpression(pattern: "(?:.*['‘](.{4})['’])?(?:.*?(-?[0-9]+) *[+]([0-9]+)?)?", options: [])
        let result = regex.firstMatch(in: label, options: [], range: NSRange(location: 0, length: label.count))
        if let nsr = result?.range(at: 2), let r = Range(nsr, in: label), let i = Int(label[r]) {
            offset = i
        }
        if let nsr = result?.range(at: 3), let r = Range(nsr, in: label), let i = Int(label[r]) {
            range = offset...(offset + i)
        }

        try super.configure()
        width = 240
        fixedMap = cases

        // Setting the resType will load the cases so we need to do this last
        if let nsr = result?.range(at: 1), let r = Range(nsr, in: label) {
            resType = String(label[r])
        } else {
            // See if we can bind to a preceding TNAM field
            guard let tnam = parentList.previous(ofType: "TNAM") else {
                throw TemplateError.invalidStructure(self, "Could not determine resource type.")
            }
            self.bind(NSBindingName("resType"), to: tnam, withKeyPath: "value")
        }
    }

    override func configure(view: NSView) {
        self.configureComboBox(view: view)
    }

    override func readData(from reader: BinaryDataReader) throws {
        tValue = try reader.read()
    }

    override func writeData(to writer: BinaryDataWriter) {
        writer.write(tValue)
    }

    override var formatter: Formatter {
        return self.sharedFormatter("INT\(T.bitWidth)") { IntFormatter<T>() }
    }

    private func loadCases() {
        cases = fixedMap
        let resources = parentList.controller.resources(ofType: resType)
        for resource in resources where range?.contains(resource.id) != false {
            let resID = resource.id - offset
            let idDisplay = self.resIDDisplay(resID)
            let caseEl = ElementCASE(value: resource.id,
                                     displayLabel: "\(idDisplay) = \(resource.name)",
                                     displayValue: "\(resource.name) = \(idDisplay)")
            cases[resID] = caseEl
        }
        value = value as NSNumber // Trigger refresh
    }

    private func resIDDisplay(_ resID: Int) -> String {
        let id = resID + offset
        if offset != 0 && range?.contains(id) != false {
            // If an offset is used, the value will be displayed as "value (#actual id)"
            return "\(resID) (#\(id))"
        }
        return String(resID)
    }

    private func updateLinkIcon() {
        guard !resType.isEmpty else {
            return
        }
        let id = Int(tValue) + offset
        if range?.contains(id) == false {
            // If out of range, show no icon
            linkIcon = nil
        } else if cases[tValue]?.label.isEmpty == true {
            // If resource exists in case list (empty label implies a resource), show link icon
            linkIcon = NSImage.followLinkFreestandingTemplateName
        } else if parentList.controller.manager.findResource(type: .init(resType), id: id, currentDocumentOnly: false) != nil {
            // If found in directory (as a last resort), show link icon
            linkIcon = NSImage.followLinkFreestandingTemplateName
        } else {
            // Resource doesn't exist, show add icon
            linkIcon = NSImage.touchBarAddDetailTemplateName
        }
    }

    func followLink(_ sender: Any) {
        let id = Int(tValue) + offset
        parentList.controller.openOrCreateResource(typeCode: resType, id: id) { [weak self] resource, isNew in
            guard let self else { return }
            // If this is new resource with a valid id, reload the cases
            let resID = resource.id - self.offset
            if isNew && self.range?.contains(resID) != false {
                self.loadCases()
                // Check if the value changed
                if resource.id != id {
                    self.value = resID as NSNumber
                    self.parentList.controller.itemValueUpdated(sender)
                }
            }
        }
    }

    override func transformedValue(_ value: Any?) -> Any? {
        return cases[value as! AnyHashable]?.displayLabel ?? self.resIDDisplay(value as! Int)
    }

    override func reverseTransformedValue(_ value: Any?) -> Any? {
        var value = value as? String
        if offset != 0 {
            value = value?.components(separatedBy: " (#").first
        }
        return super.reverseTransformedValue(value) ?? ""
    }
}
